File: techie-delight/strings/longest_substring.c
#include <stdio.h>
#include <string.h>
#include <stdlib.h>

int get_sub_string_len(char * s) {
    int s_len = strlen(s);
    char *seen = (char *) malloc(s_len);
    int i = 0;
    printf(" %s: len: %d\n",__FUNCTION__,s_len);
    for (i = 0; i < s_len; i++) {
        printf(" %s: %d\n",__FUNCTION__,__LINE__);
        if (i == 0) {
            seen[i] = s[i];
            continue;
        }
        if (s[i] != seen[i]) {
            seen[i] = s[i];
        } else {
            seen[i] = '\0';
            break;
        } 
    } 

    return strlen(seen);
}

int longest_substring(char *s) {
    int max_=0, max_so_far=0;
    int s_len = strlen(s);
    printf(" %s: len: %d\n",__FUNCTION__,s_len);
    for (int i=0; i < s_len; i++) {
        max_ = get_sub_string_len(s);
        if (max_ > max_so_far) {
            max_so_far = max_;
        }
    }

    return max_so_far;
}

int main() {
    char *s = "abcabcabcbb";

    printf(" input string is : %s\n", s);
    printf(" longest substring len is %d\n", longest_substring(s));

    return 0;
}

File: techie-delight/strings/longest_substring.py
#/usr/bin/python

# Given a string , find longest substring with non repetitive characters"
# 
# Examples:
#
# string = "abcabcbb", output = "abc", length = 3
# string = "bbbbbbb", output = "b", length = 1
# string = "pwwkew", output ="wke", length = 3


def get_substrings(s, seen_list):
    seen = ""
    if s is "":
        seen_list.append(s)
        return

    if len(s) == 1:
        seen_list.append(s)
        s = ""
        return

    for i in range(0,len(s)):
        if s[i] not in seen:
            seen = seen + s[i] 
        else:
            seen_list.append(seen)
            s = s[i:]
            get_substrings(s,seen_list)
            return        
            
def longest_substring(s):
    seen_list = []
    get_substrings(s, seen_list)

    max_ = 0
    max_so_far = 0
    max_index = 0

    for i in range(0,len(seen_list)): 
        max_ = len(seen_list[i])
        if max_ > max_so_far:
            max_so_far = max_
            max_index = i 
    
    print "string " + s + " longest substring is " + seen_list[max_index] + " of length " + str(len(seen_list[max_index])) 



s = "pwwkew"
longest_substring(s)
s = "abcabcbb"
longest_substring(s)
s = "bbbbbba"
longest_substring(s)
